### ISSUE TITLE
Make debug logs available through `adb logcat` in release builds

### ISSUE THREAD

**Issue by dkasak** [element-hq/element-x-android#4011]
#### What would you like to do?

Have debug logs available through `adb logcat` in release builds.

#### Why would you like to do it?

To make issues in production easier to debug, since logcat is far easier to grep and manipulate than files. The logs are already available through in-app retrievable log files, but this is much clunkier and harder to use.

#### How would you like to achieve it?

[Ganfra informed me](https://matrix.to/#/!LiSLXinTDCsepePiYW:matrix.org/$1xT0c7_oLkFRcUvGHGdhNX-X2O3Fo39mKhezI1-FqJg?via=matrix.org&via=element.io&via=jki.re) that currently the availability of logs over logcat is tied to the `debuggable` setting. Since release builds are not debuggable, no logs currently get sent to logcat. However, this relationship is not inherently require and we could in fact make the availability of logs through logcat not be tied to the `debuggable` setting.

### Additional context

For comparison, Signal has pretty detailed logcat logs in production builds which makes it far easier to debug issues.

I'm setting "no" to the question of providing a PR myself. This is mostly because I'm not familiar with what exactly needs to be done and I'm assuming EXA devs know much better than me. However I'm not completely closed off to researching and providing a PR if that's what's required to make it a reality.

### Are you willing to provide a PR?

No

**Comment by jmartinesp** [comment]
Just a suggestion: maybe we could enable these through some settings? The logs go through the Rust FFI layer so they can affect performance a bit, and not having them enabled by default could mitigate some mistakes that could end up displaying sensitive info in the logs (they'd still be saved to disk, but accessing those files is harder).

**Comment by spaetz** [comment]
> who are we defending against?

Performance loss :-) SCNR.

**Comment by jmartinesp** [comment]
> So the files are currently created directly on the Rust layer and they never go through FFI?

Technically the text is sent to Rust through the FFI layer and written in the log files by Rust, and if the `write_to_stdout_or_system` flag is enabled (as it happens in debug and nightly builds) it calls a C function to write the text to the logcat, so there may not be any noticeable performance loss at all.

> IMO, we have to up our game and never let sensitive info end up in the logs by mistake

Yep, that's why I said it might be a mitigation, since it would only make it more difficult for an attacker to get this data, but not impossible.

The proposed toggle might not make sense given the two points above, so just enabling the logs would be alright.

**Comment by bmarty** [comment]
I do not agree with this change. In the past, we had several cases of application logging critical data by mistake. This is happening. And we (application code) do not control all the logs. The SDK is also logging lots of data. Printing the log in production may let any third party applications with `READ_LOGS` permission will be able to access this logs. If a future leak is discovered, the severity will be higher because of this.

It's also sort of a bad practice for Android application to print log in production build. And it is bad for performance, even if nowadays this will be hardly noticeable.

More info about this in https://developer.android.com/privacy-and-security/risks/log-info-disclosure#overview

If you want to debug the application for a specific issue, just use a debug build.

**Comment by bmarty** [comment]
After internal discussion, we will add a feature flag (disabled by default) to enable outputting logs to logcat in production build. Note that using `adb logcat` may not make those log visible because the application is not `debuggable` and some filtering may occurs.

**Comment by dkasak** [comment]
> In the past, we had several cases of application logging critical data by mistake.

The way to fix this problem is by ensuring the application doesn't log critical data by mistake. If we log critical data (passwords, cryptographic keys and secrets), that data still ends up in Element's hands when they submit a bug (!) which is a *huge* problem. We really need to solve this problem at the root rather than trying to patch over the impact.

> Printing the log in production may let any third party applications with READ_LOGS permission will be able to access this logs.

It's been a long time since `READ_LOGS` was available for end-user applications. Only a handful of firmware system apps have this permission on my phone, for example.

> If you want to debug the application for a specific issue, just use a debug build.

This is unrealistic. Often you'll stumble on a hard to reproduce issue in production and you have no way of simply replicating it on a debug build. This isn't a theoretical concern, I encountered this situation many times.

> , we will add a feature flag (disabled by default) to enable outputting logs to logcat in production build

What do we mean by a "feature flag"? A flag toggleable in the dev settings (after the 7 tap secret handshake)?

**Comment by jmartinesp** [comment]
> The way to fix this problem is by ensuring the application doesn't log critical data by mistake. If we log critical data (passwords, cryptographic keys and secrets), that data still ends up in Element's hands when they submit a bug (!) which is a huge problem. We really need to solve this problem at the root rather than trying to patch over the impact.

I think there are already some checks in the Rust SDK for preventing logging sensitive data, and we actively avoid adding those logs in the app, although there are some other dependencies as EC that could maybe log something accidentally (I assume they too are careful about their logs though). IMO this is not a replacement with being careful about logging sensitive data, it's just an additional step to make sure if we somehow still manage to leak something to the logs, its exposure is more limited.

> This is unrealistic. Often you'll stumble on a hard to reproduce issue in production and you have no way of simply replicating it on a debug build. This isn't a theoretical concern, I encountered this situation many times.

I agree, there are sometimes where logging into another session in a debug build will make triggering the same issue impossible. You'd still be able to upload and export logs though, so the only problem would be live debugging the issues (I know, in some cases this can be a major problem).

> What do we mean by a "feature flag"? A flag toggleable in the dev settings (after the 7 tap secret handshake)?

Yep, you go to developer settings and enable some check that has some warning like "this will make logs for the app readable through the device logs and may affect performance" , then restart the app so the new settings for tracing are applied.

**Comment by dkasak** [comment]
> IMO this is not a replacement with being careful about logging sensitive data, it's just an additional step to make sure if we 
somehow still manage to leak something to the logs, its exposure is more limited.

> Yep, you go to developer settings and enable some check that has some warning like "this will make logs for the app readable through the device logs and may affect performance" , then restart the app so the new settings for tracing are applied.

I agree with the above. As long as you can deliberately enable logcat output in production builds via a developer setting, I'm happy!
